EXTENSIONS.md

```diff
@@ -0,0 +1,113 @@
+# Linkerd Extensions
+
+Linkerd has an extension model which allows 3rd parties to add functionality.
+Each extension consists of a binary CLI named `linkerd-name` (where `name` is
+the name of the extension) as well as a set of Kubernetes resources.  To invoke
+an extension, users can run `linkerd name` which will search the current PATH
+for an executable named `linkerd-name` and execute it.
+
+## Installing Extensions
+
+To install an extension, first download the extension executable and put it
+in your PATH. The extension can then be installed into your Kubernetes cluster
+by running `linkerd <extension name> install | kubectl apply -f -`. Similarly,
+it can be uninstalled by running
+`linkerd <extension name> uninstall | kubectl delete -f -`.
+
+A full list of installed extensions can be printed by running `linkerd check`.
+
+## Developing Extensions
+
+The extension must be an executable file named `linkerd-name` where `name` is
+the name of the extension. The name must not be any of the built-in Linkerd
+commands (e.g. `check`) or extensions (e.g. `viz`).
+
+The extension must accept the following flags and respect them any time that
+it communicates with the Kubernetes API.  All of these flags must be accepted
+but may be ignored if they are not applicable.
```

Comment:
i like that these flags are recommended but not required. we could consider a similar pattern to `kubectl plugin`, with provided libraries, or, we could even consider re-using the `kubectl` libraries/examples:
https://krew.sigs.k8s.io/docs/developer-guide/develop/best-practices/#kubectl-options
https://github.com/kubernetes/cli-runtime/

also, how do we expect `linkerd` flags to interact with extension flags? what if a user sets two different `--context` values? for comparison, `kubectl` disallows all flags on itself when running plugins:
```bash
$ kubectl --context foo access-matrix
Error: flags cannot be placed before plugin name: --context
```

Reply:
I don't think that it's possible to set the same flag to different values on the same command.  Conceptually, the position of the flag is not important: `kubectl --context foo access-matrix` is semantically equivalent to `kubectl access-matrix --context foo`.  I believe that kubectl requires that flags come after the plugin name because of the way they do argument parsing.  It's fairly likely that we would do argument parsing the same way and therefore impose the same restriction on flag ordering.